File: train/src/models/layers/Residual.lua
local nnlib = require('cudnn')
local conv = nnlib.SpatialConvolution
local batchnorm = nnlib.SpatialBatchNormalization
local relu = nnlib.ReLU

-- Main convolutional block
local function convBlock(numIn,numOut,stride,type)
    local s = nn.Sequential()
    if type ~= 'no_preact' then
        s:add(batchnorm(numIn))
        s:add(relu(true))        
    end
    s:add(conv(numIn,numOut/2,1,1))
    s:add(batchnorm(numOut/2))
    s:add(relu(true))
    s:add(conv(numOut/2,numOut/2,3,3,stride,stride,1,1))
    s:add(batchnorm(numOut/2))
    s:add(relu(true))
    s:add(conv(numOut/2,numOut,1,1))
    return s
end

-- Skip layer
local function skipLayer(numIn,numOut,stride, useConv)
    if useConv then
        return nn.Sequential()
            :add(batchnorm(numIn))
            :add(relu(true))        
            :add(conv(numIn,numOut,1,1,stride,stride))
    end

    if numIn == numOut and stride == 1 then
        return nn.Identity()
    else
        return nn.Sequential()
            :add(batchnorm(numIn))
            :add(relu(true))        
            :add(conv(numIn,numOut,1,1,stride,stride))
    end
end

-- Residual block
function Residual(numIn,numOut,stride,type, useConv)
    local stride = stride or 1
    local type = type or 'preact'
    local useConv = useConv or false
    return nn.Sequential()
        :add(nn.ConcatTable()
            :add(convBlock(numIn,numOut,stride,type))
            :add(skipLayer(numIn,numOut,stride,useConv)))
        :add(nn.CAddTable(true))
end

return Residual

File: SPPE/src/models/layers/Residual.py
import torch.nn as nn
import math
from .util_models import ConcatTable, CaddTable, Identity
from opt import opt


def Residual(numIn, numOut, *arg, stride=1, net_type='preact', useConv=False, **kw):
    con = ConcatTable([convBlock(numIn, numOut, stride, net_type),
                       skipLayer(numIn, numOut, stride, useConv)])
    cadd = CaddTable(True)
    return nn.Sequential(con, cadd)


def convBlock(numIn, numOut, stride, net_type):
    s_list = []
    if net_type != 'no_preact':
        s_list.append(nn.BatchNorm2d(numIn))
        s_list.append(nn.ReLU(True))

    conv1 = nn.Conv2d(numIn, numOut // 2, kernel_size=1)
    if opt.init:
        nn.init.xavier_normal(conv1.weight, gain=math.sqrt(1 / 2))
    s_list.append(conv1)

    s_list.append(nn.BatchNorm2d(numOut // 2))
    s_list.append(nn.ReLU(True))

    conv2 = nn.Conv2d(numOut // 2, numOut // 2, kernel_size=3, stride=stride, padding=1)
    if opt.init:
        nn.init.xavier_normal(conv2.weight)
    s_list.append(conv2)
    s_list.append(nn.BatchNorm2d(numOut // 2))
    s_list.append(nn.ReLU(True))

    conv3 = nn.Conv2d(numOut // 2, numOut, kernel_size=1)
    if opt.init:
        nn.init.xavier_normal(conv3.weight)
    s_list.append(conv3)

    return nn.Sequential(*s_list)


def skipLayer(numIn, numOut, stride, useConv):
    if numIn == numOut and stride == 1 and not useConv:
        return Identity()
    else:
        conv1 = nn.Conv2d(numIn, numOut, kernel_size=1, stride=stride)
        if opt.init:
            nn.init.xavier_normal(conv1.weight, gain=math.sqrt(1 / 2))
        return nn.Sequential(
            nn.BatchNorm2d(numIn),
            nn.ReLU(True),
            conv1
        )
